### PR TITLE
Feature - Remove the restrictive validations for additional contract files.

### PR DESCRIPTION
**Pull request by CheesyBoy123**
Motivation:

Schema validations for additional schemas was to strict. These additional schemas may only be JSON Schemas instead of openapi schemas.

See the related issues:
https://github.com/eclipse-vertx/vertx-openapi/issues/77
https://github.com/eclipse-vertx/vertx-openapi/issues/16

Conformance:

You should have signed the Eclipse Contributor Agreement as explained in https://github.com/eclipse/vert.x/blob/master/CONTRIBUTING.md
Please also make sure you adhere to the code style guidelines: https://github.com/vert-x3/wiki/wiki/Vert.x-code-style-guidelines

### CODE REVIEW

File: src/main/java/io/vertx/openapi/contract/OpenAPIVersion.java [hunk 1]
```diff
@@ -75,16 +76,42 @@ public static OpenAPIVersion fromContract(JsonObject contract) {
...
+  /**
+   * Validates additional contract files against the openapi schema. If validations fails, try to validate against the
+   * json schema specifications only.
+   *
+   * @param vertx                   The related Vert.x instance.
+   * @param repo                    The SchemaRepository to do the validations with.
+   * @param file                    The additional json contract to validate.
+   */
+  public Future<Void> validateAdditionalContractFiles(Vertx vertx, SchemaRepository repo, JsonObject file) {
+    return validateContract(vertx, repo, file)
```
Comment (pk-work) [inline src/main/java/io/vertx/openapi/contract/OpenAPIVersion.java:92]: Why did this method also validate the contract?

File: src/main/java/io/vertx/openapi/contract/OpenAPIVersion.java [hunk 2]
```diff
@@ -75,16 +76,42 @@ public static OpenAPIVersion fromContract(JsonObject contract) {
...
+      .compose(this::checkOutputUnit)
+      .recover((t) -> {
+        //Schema validation failed using the openapi spec, so instead try to use json schema spec only.
+        return checkOutputUnit(repo.validator(draft.getIdentifier()).validate(file));
```
Comment (pk-work) [inline src/main/java/io/vertx/openapi/contract/OpenAPIVersion.java:96]: We should run this valiation in executeBlocking, because it could block the EL

File: src/main/java/io/vertx/openapi/contract/OpenAPIVersion.java [hunk 3]
```diff
@@ -75,16 +76,42 @@ public static OpenAPIVersion fromContract(JsonObject contract) {
...
+
+  /**
+   * Validates additional contract files against the openapi schema. If validations fails, try to validate against the
+   * json schema specifications only.
+   *
+   * @param vertx                   The related Vert.x instance.
+   * @param repo                    The SchemaRepository to do the validations with.
+   * @param file                    The additional json contract to validate.
+   */
+  public Future<Void> validateAdditionalContractFiles(Vertx vertx, SchemaRepository repo, JsonObject file) {
```
Comment (pk-work) [inline src/main/java/io/vertx/openapi/contract/OpenAPIVersion.java:91]: In my opinion `validateAdditionalContractFiles` is missleading, as it validates only a single file. we should name it `validateAdditionalContractFile`. Or pass all files.

This method should have its own unit test in `OpenAPIVersionTest`. The `testValidate` method in `OpenAPIVersionTest` should get renamed in `validateContractTest` and the new test method should be called `validateAdditionalContractFileTest`.

File: src/test/resources/io/vertx/openapi/contract/split/malformedComponents.json [hunk 4]
```diff
@@ -0,0 +1,24 @@
+{
+  "title": "my json api",
+  "description": "taken from https://stackoverflow.com/questions/16826128/why-is-this-json-schema-invalid-using-any-type",
```
Comment (pk-work) [inline src/test/resources/io/vertx/openapi/contract/split/malformedComponents.json:3]: Do you know the license of the Stackoverflow example? If not we should create our own to avoid legal issues.

File: src/test/java/io/vertx/openapi/contract/OpenAPIContractTest.java [hunk 5]
```diff
@@ -112,6 +114,48 @@ void testFromFailsInvalidSpec(Vertx vertx, VertxTestContext testContext) {
...
+
+    Map<String, JsonObject> additionalMalformedSpecFiles = ImmutableMap.of("https://example.com/petstore", malformedComponents);
+
+    OpenAPIContract.from(vertx, contract.copy(), additionalMalformedSpecFiles)
+      .onComplete(handler -> testContext.verify(() -> {
+        assertTrue(handler.failed());
+        assertThat(handler.cause()).isInstanceOf(OpenAPIContractException.class);
+        assertThat(handler.cause()).hasMessageThat()
+          .isEqualTo("The passed OpenAPI contract is invalid: Found issue in specification for reference: " +
+            "Instance type string is invalid. Expected array");
```
Comment (pk-work) [inline src/test/java/io/vertx/openapi/contract/OpenAPIContractTest.java:154]: In the linked stackoverflow example there is an issue with type "any". How is this error message related to this error? Do we test the correct error?

**Review by pk-work — CHANGES_REQUESTED** [review]

File: src/main/java/io/vertx/openapi/contract/OpenAPIVersion.java (hunk 2 above)
Comment (CheesyBoy123) [reply src/main/java/io/vertx/openapi/contract/OpenAPIVersion.java:96]: Good point, i've updated this to be in an executeBlocking().

File: src/main/java/io/vertx/openapi/contract/OpenAPIVersion.java (hunk 1 above)
Comment (CheesyBoy123) [reply src/main/java/io/vertx/openapi/contract/OpenAPIVersion.java:92]: It first tries to validate it against the openapi scheme first. Maybe this is too much validations and we can just have it validate against the OpenApi version's draft schemes instead?

File: src/test/resources/io/vertx/openapi/contract/split/malformedComponents.json (hunk 4 above)
Comment (CheesyBoy123) [reply src/test/resources/io/vertx/openapi/contract/split/malformedComponents.json:3]: i've removed this example and replaced with my own.

File: src/test/java/io/vertx/openapi/contract/OpenAPIContractTest.java (hunk 5 above)
Comment (CheesyBoy123) [reply src/test/java/io/vertx/openapi/contract/OpenAPIContractTest.java:154]: I've removed this example to replace it with a more straight forward example.

File: src/main/java/io/vertx/openapi/contract/OpenAPIVersion.java (hunk 1 above)
Comment (CheesyBoy123) [reply src/main/java/io/vertx/openapi/contract/OpenAPIVersion.java:92]: I've decided to remove validation against the openapi spec as that is really just a super set of the json schema spec. The additional files only really need to be a valid json spec.

File: src/main/java/io/vertx/openapi/contract/OpenAPIVersion.java (hunk 3 above)
Comment (CheesyBoy123) [reply src/main/java/io/vertx/openapi/contract/OpenAPIVersion.java]: Updated these and added an additional test for validateAdditionalContractFile itself.

File: src/test/java/io/vertx/openapi/contract/OpenAPIContractTest.java [hunk 6]
```diff
@@ -112,6 +114,48 @@ void testFromFailsInvalidSpec(Vertx vertx, VertxTestContext testContext) {
...
+        .onSuccess(splitResp -> testContext.verify(() -> {
+          assertThat(splitResp.getRawContract()).isEqualTo(validJsonRef);
+          testContext.completeNow();
+        }))
+        .onFailure(testContext::failNow))
+      .onFailure(testContext::failNow);
+  }
+
+  @Test
+  @Timeout(value = 2, timeUnit = TimeUnit.HOURS)
```
Comment (pk-work) [inline src/test/java/io/vertx/openapi/contract/OpenAPIContractTest.java:140]:
```suggestion
  @Timeout(value = 2, timeUnit = TimeUnit.SECONDS)
```

File: src/test/java/io/vertx/openapi/contract/OpenAPIContractTest.java [hunk 7]
```diff
@@ -124,12 +168,29 @@ void testInvalidAdditionalSpecFiles(Vertx vertx, VertxTestContext testContext) {
...
-          "The passed OpenAPI contract is invalid: Found issue in specification for reference: https://example.com/petstore";
+          "The passed OpenAPI contract is invalid: Found issue in specification for reference: " +
+            "Can't resolve 'https://example.com/petstore#/components/schemas/Pet', only internal refs are supported.";
         assertThat(t).hasMessageThat().isEqualTo(expectedErrorMessage);
         testContext.completeNow();
       })));
   }
 
+  @Test
+  @Timeout(value = 2, timeUnit = TimeUnit.HOURS)
```
Comment (pk-work) [inline src/test/java/io/vertx/openapi/contract/OpenAPIContractTest.java:179]:
```suggestion
  @Timeout(value = 2, timeUnit = TimeUnit.SECONDS)
```

File: src/test/java/io/vertx/openapi/contract/OpenAPIVersionTest.java [hunk 8]
```diff
@@ -30,16 +32,19 @@
...
+import java.util.concurrent.TimeUnit;
 import java.util.function.Consumer;
 import java.util.function.Function;
 import java.util.stream.Collectors;
 import java.util.stream.Stream;
 
 import static com.google.common.truth.Truth.assertThat;
-import static io.vertx.openapi.ResourceHelper.TEST_RESOURCE_PATH;
-import static io.vertx.openapi.impl.Utils.EMPTY_JSON_OBJECT;
+import static io.vertx.openapi.ResourceHelper.*;
```
Comment (pk-work) [inline src/test/java/io/vertx/openapi/contract/OpenAPIVersionTest.java:44]: I'm personally not a friend of star imports, I'd appreciate if we can avoid them.

**Review by pk-work — CHANGES_REQUESTED** [review]

File: src/test/java/io/vertx/openapi/contract/OpenAPIVersionTest.java (hunk 8 above)
Comment (CheesyBoy123) [reply src/test/java/io/vertx/openapi/contract/OpenAPIVersionTest.java:44]: Intellij automatically changing these to * imports. I'll have to update my settings 😄

File: src/test/java/io/vertx/openapi/contract/OpenAPIContractTest.java (hunk 7 above)
Comment (CheesyBoy123) [reply src/test/java/io/vertx/openapi/contract/OpenAPIContractTest.java:179]: oops, remnants of my debugging 😆

**Review by pk-work — APPROVED** [review]
